histomicstk/segmentation/label/trace_boundaries_opt.cpp

```diff
@@ -0,0 +1,588 @@
+/*
+C++ version of trace_boundaries_opt
+*/
+#include "trace_boundaries_opt.h"
+
+#include <iostream>
+#include <list>
+#include <cmath>
+#include <vector>
+#include <set>
+#include <algorithm>
+
+struct Points {
+    int x, y;
+};
+
+void rot90(int nrows, int ncols,
+  std::vector <std::vector<int> > input,
+  std::vector <std::vector<int> > &output)
+{
+    for (int i=0; i<nrows; i++){
+      for (int j=0;j<ncols; j++){
+        output[j][nrows-1-i] = input[i][j];
+      }
+    }
+}
+
+std::vector <std::vector<std::vector<int> > > trace_boundary(std::vector <std::vector<int> > imLabels, int connectivity)
+{
+    std::vector <std::vector<std::vector<int> > > output;
+
+    // start ----- rotate matrix
+
+    int nrows_img = imLabels.size();
+    int ncols_img = imLabels[0].size();
+
+    // initialize matrix for 90, 180, 270 degrees
+    std::vector <std::vector<int> > imLabels_90(ncols_img, std::vector<int>(nrows_img));
+    std::vector <std::vector<int> > imLabels_180(nrows_img, std::vector<int>(ncols_img));
+    std::vector <std::vector<int> > imLabels_270(ncols_img, std::vector<int>(nrows_img));
+
+    // rotate label matrix for 90, 180, 270 degrees
+    rot90(nrows_img, ncols_img, imLabels, imLabels_270);
+    rot90(ncols_img, nrows_img, imLabels_270, imLabels_180);
+    rot90(nrows_img, ncols_img, imLabels_180, imLabels_90);
+
+    // end ----- rotate matrix
+
+
+    // start ----- find unique id of labels
+
+    // 2d to 1d array
+    std::vector<int> im1D;
+    im1D.reserve(nrows_img * ncols_img);
+    for(int i = 0; i < nrows_img; i++) {
+       const std::vector<int>& v = imLabels[i];
+       im1D.insert(im1D.end(), v.begin(), v.end());
+    }
```

Comment:
The creation of im1D is not needed to generate label_set.

Instead you can just do an iterator on imLabel and insert the elements into label_set directly.